src/controller.ts

```diff
@@ -1,12 +1,90 @@
-// tslint:disable-next-line:no-empty-interface
-export interface IController {}
+import Optional from "optional.js";
+import { EventHandler, Failure, makePromise, Resolver, Success } from "./events";
+import { INativeView } from "./nativeview";
+import Navigation from "./navigation";
 
-export interface ControllerCreator<T extends IController> {
-    new(...args: any[]): T;
+export interface IController {
+    readonly view: INativeView;
+    getValues(fields: string[]): Promise<string[]>;
+    render(model?: string);
+    renderComponent(name: string, model?: string);
+    display();
+    onBind(): Map<string, EventHandler>;
+    onFinish();
+    onPageCreated();
+    onPageRestart();
+    onPageVisible();
 }
 
-export function createController<T extends IController>(
-    ctor: ControllerCreator<T>,
-    ...args: any[]): T {
-    return new ctor(...args);
+export class BaseController implements IController {
+    public readonly view: INativeView;
+    constructor(view: INativeView, ...args: any[]) {
+        this.view = view;
+    }
+
+    public getValues(fields: string[]): Promise<string[]> {
+        return makePromise((success: Success<string[]>, failure: Failure) => {
```

Comment:
Couple of questions.
1. What is the value of failure?
2. What if JSON.parse fails, we should be returning failure rather than Success I believe.
3. As this seems to be the contract between native and Calatrava; why do you think wrapping it in a promise is important?

Reply:
Failure is of type (error: Error) => void.